File: internal/documents/documents.go
// Package documents is the tenant-content document API (SPEC-07 §2, FR-SRC-02,
// FR-ADM-03). Documents, their versions and their chunks are TENANT content
// (schemas/tenant.sql), never control-plane data (C-3): every read and the soft
// delete reach the tenant database ONLY through a *tenant.DB obtained from the
// Resolver (ADR-0003) — there is no tenant_id column and no control-plane path
// to this data. The tenant is always the one resolved from the authenticated
// principal (FR-ACC-03), never a request parameter.
//
// The ingest job it enqueues (ingest_document) is control-plane queue state, so
// that one write goes to the control-plane jobs table via a JobEnqueuer; the raw
// upload bytes go to object storage via the Storage seam (EPIC-06).
//
// The WRITE side of this same tenant content — building a version and its chunks
// and flipping documents.current_version atomically — is TenantStore.Put (put.go,
// STORY-05.1, ADR-0008/ADR-0033): given a fully parsed, chunked and embedded
// version, it persists it in one transaction so a query never sees a half-built
// document. The upstream pipeline stages that produce a PutInput (parser, chunker,
// embedder) and the sink orchestration/worker are EPIC-05.2+ and EPIC-09; the
// upload connector is EPIC-06.
package documents

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"time"
)

const (
	defaultListLimit = 50
	maxListLimit     = 200
	// defaultMaxUploadBytes is the FR-SRC-02 default upload ceiling (50 MB). It is
	// configurable (config.MaxUploadBytes); this is the fallback floor.
	defaultMaxUploadBytes int64 = 50 << 20
)

// Document is the API view of a tenant document (schemas/tenant.sql documents).
// It is the list-row projection; Get returns a DocumentDetail that adds the
// current version metadata.
type Document struct {
	ID             string          `json:"id"`
	SourceID       string          `json:"source_id"`
	ExternalID     string          `json:"external_id"`
	Title          *string         `json:"title,omitempty"`
	URI            *string         `json:"uri,omitempty"`
	MimeType       *string         `json:"mime_type,omitempty"`
	Status         string          `json:"status"`
	CurrentVersion *string         `json:"current_version,omitempty"`
	Metadata       json.RawMessage `json:"metadata,omitempty"`
	FirstSeenAt    time.Time       `json:"first_seen_at"`
	LastSeenAt     time.Time       `json:"last_seen_at"`
	DeletedAt      *time.Time      `json:"deleted_at,omitempty"`
}

// VersionMeta is the current-version metadata returned with a document. The full
// normalised text (document_versions.content) is included only when the caller
// asks (?content=true), keeping the default Get response small.
type VersionMeta struct {
	ID          string    `json:"id"`
	ContentHash string    `json:"content_hash"` // hex-encoded sha256
	CharCount   int       `json:"char_count"`
	Parser      *string   `json:"parser,omitempty"`
	CreatedAt   time.Time `json:"created_at"`
	Content     *string   `json:"content,omitempty"` // only when ?content=true
}

// DocumentDetail is GET /v1/documents/{id}: the document plus its current version
// metadata (SPEC-07 §2). CurrentVersion is nil when the document has no built
// version yet (should not happen for an active document — SPEC-03 §2 invariant 1).
type DocumentDetail struct {
	Document
	CurrentVersion *VersionMeta `json:"current_version_meta,omitempty"`
}

// Chunk is the debugging view of a chunk (FR-ADM-03). The embedding vector is
// deliberately never returned: it is large, opaque, and not useful to a human
// debugging retrieval.
type Chunk struct {
	ID             string          `json:"id"`
	Position       int             `json:"position"`
	HeadingPath    []string        `json:"heading_path"`
	Content        string          `json:"content"`
	TokenCount     int             `json:"token_count"`
	EmbeddingModel string          `json:"embedding_model"`
	Metadata       json.RawMessage `json:"metadata,omitempty"`
	CreatedAt      time.Time       `json:"created_at"`
}

// Page is the SPEC-07 §1 pagination envelope for documents.
type Page struct {
	Items      []Document `json:"items"`
	NextCursor string     `json:"next_cursor,omitempty"`
}

// ChunkPage is the SPEC-07 §1 pagination envelope for a document's chunks.
type ChunkPage struct {
	Items      []Chunk `json:"items"`
	NextCursor string  `json:"next_cursor,omitempty"`
}

// ListFilter selects documents by source, status and a free-text query (SPEC-07
// §2: "filter by source, status, q"). An empty field is "no filter".
type ListFilter struct {
	SourceID string
	Status   string
	Q        string
}

// Cursor is the opaque keyset position for document List pagination: the
// (first_seen_at, id) of the last returned row.
type Cursor struct {
	FirstSeenAt time.Time `json:"f"`
	ID          string    `json:"i"`
}

// ChunkCursor is the keyset position for chunk pagination: the position of the
// last returned chunk (chunks are ordered by position within the current version).
type ChunkCursor struct {
	Position int `json:"p"`
}

// encodeCursor serialises a Cursor to an opaque base64url token.
func encodeCursor(c Cursor) string {
	b, _ := json.Marshal(c)
	return base64.RawURLEncoding.EncodeToString(b)
}

// decodeCursor parses a base64url document cursor token.
func decodeCursor(s string) (*Cursor, error) {
	b, err := base64.RawURLEncoding.DecodeString(s)
	if err != nil {
		return nil, err
	}
	var c Cursor
	if err := json.Unmarshal(b, &c); err != nil {
		return nil, err
	}
	return &c, nil
}

// encodeChunkCursor / decodeChunkCursor are the chunk-page equivalents.
func encodeChunkCursor(c ChunkCursor) string {
	b, _ := json.Marshal(c)
	return base64.RawURLEncoding.EncodeToString(b)
}

func decodeChunkCursor(s string) (*ChunkCursor, error) {
	b, err := base64.RawURLEncoding.DecodeString(s)
	if err != nil {
		return nil, err
	}
	var c ChunkCursor
	if err := json.Unmarshal(b, &c); err != nil {
		return nil, err
	}
	return &c, nil
}

// clampLimit applies the default and maximum page sizes.
func clampLimit(n int) int {
	if n <= 0 {
		return defaultListLimit
	}
	if n > maxListLimit {
		return maxListLimit
	}
	return n
}

// validUUID is a cheap guard so a non-UUID path segment scans as ErrNotFound
// rather than raising a Postgres type error (mirrors internal/cp/sources).
func validUUID(s string) bool {
	if len(s) != 36 {
		return false
	}
	for i, c := range s {
		if i == 8 || i == 13 || i == 18 || i == 23 {
			if c != '-' {
				return false
			}
			continue
		}
		isHex := (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')
		if !isHex {
			return false
		}
	}
	return true
}

// invalid builds a ValidationError with a formatted message.
func invalid(format string, args ...any) *ValidationError {
	return &ValidationError{Msg: fmt.Sprintf(format, args...)}
}

File: web/lib/documents.ts
import { useQuery, type UseQueryResult } from "@tanstack/react-query";
import { apiFetch } from "./api";
import { useTenant } from "./tenant";

// The types mirror the control-plane JSON (internal/documents/documents.go). The
// documents browser is read-only (FR-ADM-03): list, detail with current-version
// metadata, and the chunk debug view. Upload/delete stay on the Bearer surface.
export type DocumentStatus = "active" | "deleted";

export type Document = {
  id: string;
  source_id: string;
  external_id: string;
  title?: string;
  uri?: string;
  mime_type?: string;
  status: DocumentStatus;
  current_version?: string;
  metadata?: Record<string, unknown>;
  first_seen_at: string;
  last_seen_at: string;
  deleted_at?: string;
};

// VersionMeta is the current-version metadata. `content` is present only when the
// detail was fetched with ?content=true (not used by the browser's default view).
export type VersionMeta = {
  id: string;
  content_hash: string;
  char_count: number;
  parser?: string;
  created_at: string;
  content?: string;
};

export type DocumentDetail = Document & { current_version_meta?: VersionMeta };

// Chunk is the debug view; the embedding vector is never returned by the API.
export type Chunk = {
  id: string;
  position: number;
  heading_path: string[];
  content: string;
  token_count: number;
  embedding_model: string;
  metadata?: Record<string, unknown>;
  created_at: string;
};

export type DocumentListPage = { items: Document[]; next_cursor?: string };
export type ChunkPage = { items: Chunk[]; next_cursor?: string };

// DocumentFilter carries the optional list filters (SPEC-07 §2: source, status,
// q). An empty filter lists the whole tenant, newest first.
export type DocumentFilter = {
  source?: string;
  status?: DocumentStatus;
  q?: string;
};

// ensureOk turns any non-2xx into an Error carrying the envelope's error.message
// (falling back to the status). apiFetch already throws Unauthorized on 401.
async function ensureOk(res: Response): Promise<Response> {
  if (res.ok) return res;
  let message = `request failed with status ${res.status}`;
  try {
    const body = (await res.json()) as { error?: { message?: string } };
    if (body?.error?.message) message = body.error.message;
  } catch {
    // non-JSON body — keep the status-based default
  }
  throw new Error(message);
}

function base(tenantId: string): string {
  return `/admin/tenants/${tenantId}/documents`;
}

export async function listDocuments(
  tenantId: string,
  filter: DocumentFilter = {},
  opts: { limit?: number; cursor?: string } = {},
): Promise<DocumentListPage> {
  const q = new URLSearchParams();
  if (filter.source) q.set("source", filter.source);
  if (filter.status) q.set("status", filter.status);
  if (filter.q) q.set("q", filter.q);
  if (opts.limit != null) q.set("limit", String(opts.limit));
  if (opts.cursor) q.set("cursor", opts.cursor);
  const qs = q.toString();
  const res = await ensureOk(await apiFetch(`${base(tenantId)}${qs ? `?${qs}` : ""}`));
  return (await res.json()) as DocumentListPage;
}

// listAllDocuments follows next_cursor to load EVERY matching document, so the
// admin table (which paginates client-side, usePagination) shows the true total
// instead of just the server's first page. Requests the server max (200) per round
// trip. ponytail: capped at 100 pages (20k docs); a larger corpus should move the
// table to true server-side cursor paging rather than lifting this cap.
export async function listAllDocuments(
  tenantId: string,
  filter: DocumentFilter = {},
): Promise<DocumentListPage> {
  const items: Document[] = [];
  let cursor: string | undefined;
  for (let i = 0; i < 100; i++) {
    const page = await listDocuments(tenantId, filter, { limit: 200, cursor });
    items.push(...page.items);
    if (!page.next_cursor) break;
    cursor = page.next_cursor;
  }
  return { items };
}

export async function getDocument(tenantId: string, id: string): Promise<DocumentDetail> {
  const res = await ensureOk(await apiFetch(`${base(tenantId)}/${id}`));
  return (await res.json()) as DocumentDetail;
}

export async function getChunks(
  tenantId: string,
  id: string,
  opts: { limit?: number; cursor?: string } = {},
): Promise<ChunkPage> {
  const q = new URLSearchParams();
  if (opts.limit != null) q.set("limit", String(opts.limit));
  if (opts.cursor) q.set("cursor", opts.cursor);
  const qs = q.toString();
  const res = await ensureOk(await apiFetch(`${base(tenantId)}/${id}/chunks${qs ? `?${qs}` : ""}`));
  return (await res.json()) as ChunkPage;
}

// useDocuments loads the current tenant's documents for the given filter, keyed by
// tenant and filter so switching either refetches; disabled until a tenant is known.
export function useDocuments(filter: DocumentFilter = {}): UseQueryResult<DocumentListPage, Error> {
  const tenantId = useTenant().current?.id;
  return useQuery({
    queryKey: ["documents", tenantId, filter],
    queryFn: () => listAllDocuments(tenantId as string, filter),
    enabled: !!tenantId,
  });
}

// useDocument loads one document by id, keyed by tenant and id.
export function useDocument(id: string): UseQueryResult<DocumentDetail, Error> {
  const tenantId = useTenant().current?.id;
  return useQuery({
    queryKey: ["document", tenantId, id],
    queryFn: () => getDocument(tenantId as string, id),
    enabled: !!tenantId && !!id,
  });
}

// useChunks loads one document's chunks, keyed by tenant and id.
export function useChunks(id: string): UseQueryResult<ChunkPage, Error> {
  const tenantId = useTenant().current?.id;
  return useQuery({
    queryKey: ["chunks", tenantId, id],
    queryFn: () => getChunks(tenantId as string, id),
    enabled: !!tenantId && !!id,
  });
}
